lib/IRGen/GenPack.cpp

```diff
@@ -205,60 +206,134 @@ static Address emitFixedSizeMetadataPackRef(IRGenFunction &IGF,
 
-  // Find all the pack archetypes appearing in the pattern type.
-  SmallVector<Type, 2> patternPacks;
-  patternTy->getTypeParameterPacks(patternPacks);
+  Address patternPackAddress(patternPack, IGF.IGM.TypeMetadataPtrTy,
+                             IGF.IGM.getPointerAlignment());
 
+  // Load the metadata pack element from the current source index.
+  Address fromPtr(
+      IGF.Builder.CreateInBoundsGEP(patternPackAddress.getElementType(),
+                                    patternPackAddress.getAddress(), index),
+      patternPackAddress.getElementType(), patternPackAddress.getAlignment());
+  llvm::Value *metadata = IGF.Builder.CreateLoad(fromPtr);
+
+  // Bind the metadata pack element to the element archetype.
+  IGF.setScopedLocalTypeMetadata(elementArchetype,
+                                 MetadataResponse::forComplete(metadata));
+
+  return metadata;
+}
+
+static llvm::Value *bindWitnessTableAtIndex(IRGenFunction &IGF,
+                                            CanType elementArchetype,
+                                            ProtocolConformanceRef conf,
+                                            llvm::Value *wtablePack,
+                                            llvm::Value *index) {
+  auto key = LocalTypeDataKind::forProtocolWitnessTable(conf);
+  if (auto *wtable = IGF.tryGetLocalTypeData(elementArchetype, key))
+    return wtable;
+  wtablePack =
+      IGF.Builder.CreatePointerCast(wtablePack, IGF.IGM.WitnessTablePtrPtrTy);
+
+  Address patternPackAddress(wtablePack, IGF.IGM.WitnessTablePtrTy,
+                             IGF.IGM.getPointerAlignment());
+
+  // Load the witness table pack element from the current source index.
+  Address fromPtr(
+      IGF.Builder.CreateInBoundsGEP(patternPackAddress.getElementType(),
+                                    patternPackAddress.getAddress(), index),
+      patternPackAddress.getElementType(), patternPackAddress.getAlignment());
+  auto *wtable = IGF.Builder.CreateLoad(fromPtr);
+
+  // Bind the witness table pack element to the element archetype.
+  IGF.setScopedLocalTypeData(elementArchetype, key, wtable);
+
+  return wtable;
+}
+
+static GenericEnvironment *bindElementSignatureRequirementsAtIndex(
+    IRGenFunction &IGF, CanPackExpansionType expansionTy, llvm::Value *index,
+    DynamicMetadataRequest request) {
   // Get the outer generic signature and environment.
   auto *genericEnv = cast<PackArchetypeType>(expansionTy.getCountType())
       ->getGenericEnvironment();
   auto subMap = genericEnv->getForwardingSubstitutionMap();
 
   auto genericSig = genericEnv->getGenericSignature().getCanonicalSignature();
-
   // Create an opened element signature and environment.
   auto elementSig = IGF.IGM.Context.getOpenedElementSignature(
       genericSig, expansionTy.getCountType());
   auto *elementEnv = GenericEnvironment::forOpenedElement(
       elementSig, UUID::fromTime(), expansionTy.getCountType(), subMap);
```

Comment:
Hmm.  Oh, I see, we're effectively opening a fresh environment so that we can iterate over the pack, that makes sense and probably fits into IRGen's model quite well.  It is a little unfortunate, because I think these environments are not very lightweight, but I don't think we should worry about that right now.

Did you not need to change the code we generate in `open_pack_element`?

Reply:
The opened element signature and so on are all cached, so it should be quite lightweight no?

Reply:
> Did you not need to change the code we generate in `open_pack_element`?

Though the code for that shares this code path, it looks like we do have a similar issue there.  I'll fix it in a follow-up.

Reply:
Follow-up: https://github.com/apple/swift/pull/63733